sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcUtils.scala

```diff
@@ -0,0 +1,370 @@
+      val schema = reader.getSchema
+      if (schema.getFieldNames.size == 0) {
+        None
+      } else {
+        Some(schema)
+      }
+    } catch {
+      case _: IOException => None
+    }
+  }
+
+  private[orc] def readSchema(sparkSession: SparkSession, files: Seq[FileStatus])
+      : Option[StructType] = {
+    val conf = sparkSession.sparkContext.hadoopConfiguration
+    files.map(_.getPath).flatMap(readSchema(_, conf)).headOption.map { schema =>
+      logDebug(s"Reading schema from file $files, got Hive schema string: $schema")
+      CatalystSqlParser.parseDataType(schema.toString).asInstanceOf[StructType]
+    }
+  }
+
+  private[orc] def getSchemaString(schema: StructType): String = {
+    schema.fields.map(f => s"${f.name}:${f.dataType.catalogString}").mkString("struct<", ",", ">")
+  }
+
+  private[orc] def getTypeDescription(dataType: DataType) = dataType match {
+    case st: StructType => TypeDescription.fromString(getSchemaString(st))
+    case _ => TypeDescription.fromString(dataType.catalogString)
+  }
+
+  /**
+   * Return a missing schema in a give ORC file.
+   */
+  private[orc] def getMissingSchema(
+      resolver: Resolver,
+      dataSchema: StructType,
+      partitionSchema: StructType,
+      file: Path,
+      conf: Configuration): Option[StructType] = {
+    try {
+      val fs = file.getFileSystem(conf)
+      val readerOptions = OrcFile.readerOptions(conf).filesystem(fs)
+      val reader = OrcFile.createReader(file, readerOptions)
+      val schema = reader.getSchema
+      if (schema.getFieldNames.size == 0) {
+        None
+      } else {
+        val orcSchema = if (schema.getFieldNames.asScala.forall(_.startsWith("_col"))) {
+          logInfo("Recover ORC schema with data schema")
+          var schemaString = schema.toString
+          dataSchema.zipWithIndex.foreach { case (field: StructField, index: Int) =>
+            schemaString = schemaString.replace(s"_col$index:", s"${field.name}:")
+          }
+          TypeDescription.fromString(schemaString)
+        } else {
+          schema
+        }
+
+        var missingSchema = new StructType
+        if (dataSchema.length > orcSchema.getFieldNames.size) {
+          dataSchema.filter(x => partitionSchema.getFieldIndex(x.name).isEmpty).foreach { f =>
+            if (!orcSchema.getFieldNames.asScala.exists(resolver(_, f.name))) {
+              missingSchema = missingSchema.add(f)
+            }
+          }
+        }
+        Some(missingSchema)
+      }
+    } catch {
+      case _: IOException => None
+    }
+  }
+
+  /**
+   * Return a Orc value object for the given Spark schema.
+   */
+  private[orc] def createOrcValue(dataType: DataType) =
+    OrcStruct.createValue(getTypeDescription(dataType))
+
+  /**
+   * Convert Apache ORC OrcStruct to Apache Spark InternalRow.
+   * If internalRow is not None, fill into it. Otherwise, create a SpecificInternalRow and use it.
+   */
+  private[orc] def convertOrcStructToInternalRow(
```

Comment:
like the old orc format, can we create a `OrcSerializer` to capsulate these serializing logic?

Reply:
Thanks. It's done.